src/app/models/AbstractModel.spec.ts

```diff
@@ -0,0 +1,722 @@
+        public readonly sites: Ids;
+        public readonly siteIds: Id[];
+
+        public get viewUrl(): string {
+          throw new Error("Method not implemented.");
+        }
+      }
+
+      const model = new MockModel({ siteIds: [1, 2, 3] });
+      expect(model.sites).toEqual(new Set([1, 2, 3]));
+      expect(model.siteIds).toEqual([1, 2, 3]);
+    });
+
+    it("should convert undefined", () => {
+      const model = createModel(undefined);
+      expect(model.ids).toEqual(new Set([]));
+    });
+
+    it("should convert null", () => {
+      const model = createModel(null);
+      expect(model.ids).toEqual(new Set([]));
+    });
+
+    it("should convert empty array", () => {
+      const model = createModel([]);
+      expect(model.ids).toEqual(new Set([]));
+    });
+
+    it("should convert single value array", () => {
+      const model = createModel([1]);
+      expect(model.ids).toEqual(new Set([1]));
+    });
+
+    it("should convert multi value array", () => {
+      const model = createModel([1, 2, 3]);
+      expect(model.ids).toEqual(new Set([1, 2, 3]));
+    });
+  });
+
+  describe("BawDateTime", () => {
+    function createModel(data: string) {
+      class MockModel extends AbstractModel {
+        @BawDateTime()
+        public readonly date: DateTime;
+
+        public get viewUrl(): string {
+          throw new Error("Method not implemented.");
+        }
+      }
+
+      return new MockModel(data === undefined ? {} : { date: data });
+    }
+
+    it("should handle persist option", () => {
+      class MockModel extends AbstractModel {
+        @BawDateTime({ persist: true })
+        public readonly date: DateTime;
+
+        public get viewUrl(): string {
+          throw new Error("Method not implemented.");
+        }
+      }
+
+      const model = new MockModel({ date: "2019-01-01T00:00:00" });
+      expect(model["_attributes"]).toEqual(["date"]);
+    });
+
+    it("should handle override key option", () => {
+      class MockModel extends AbstractModel {
+        @BawDateTime({ key: "timestamp" })
+        public readonly dateTime: DateTime;
+        public readonly timestamp: string;
+
+        public get viewUrl(): string {
+          throw new Error("Method not implemented.");
+        }
+      }
+
+      const model = new MockModel({ timestamp: "2019-01-01T00:00:00" });
+      expect(model.dateTime).toEqual(DateTime.fromISO("2019-01-01T00:00:00"));
+      expect(model.timestamp).toEqual("2019-01-01T00:00:00");
+    });
+
+    it("should convert undefined", () => {
+      const model = createModel(undefined);
+      expect(model.date).toEqual(null);
+    });
+
+    it("should convert null", () => {
+      const model = createModel(null);
+      expect(model.date).toEqual(null);
+    });
+
+    it("should convert timestamp string", () => {
+      const model = createModel("2019-01-01T00:00:00");
+      expect(model.date).toEqual(DateTime.fromISO("2019-01-01T00:00:00"));
+    });
+  });
+
```

Comment:
That is, to my understanding of my code, currently how the system works. The value comes in as `2020-04-27T12:20:00+10:00`, and goes back as `2020-04-27T12:20:00+10:00`. This is because the `toJSON()` method in the `AbstractModel` handles conversion of javascript types back to their value. Ie, Converting DateTime objects back to their ISO format.

Reply:
Great, add a unit test please